pkg/apis/core/validation/validation.go

```diff
@@ -4160,16 +4163,25 @@ func ValidateService(service *core.Service, allowAppProtocol bool) field.ErrorLi
 				allErrs = append(allErrs, field.Invalid(portPath, port.Port, fmt.Sprintf("may not expose port %v externally since it is used by kubelet", ports.KubeletPort)))
 			}
 		}
-		if service.Spec.ClusterIP == "None" {
-			allErrs = append(allErrs, field.Invalid(specPath.Child("clusterIP"), service.Spec.ClusterIP, "may not be set to 'None' for LoadBalancer services"))
+		if isHeadlessService(service) {
+			allErrs = append(allErrs, field.Invalid(specPath.Child("clusterIPs").Index(0), service.Spec.ClusterIPs[0], "may not be set to 'None' for LoadBalancer services"))
 		}
 	case core.ServiceTypeNodePort:
-		if service.Spec.ClusterIP == "None" {
-			allErrs = append(allErrs, field.Invalid(specPath.Child("clusterIP"), service.Spec.ClusterIP, "may not be set to 'None' for NodePort services"))
+		if isHeadlessService(service) {
+			allErrs = append(allErrs, field.Invalid(specPath.Child("clusterIPs").Index(0), service.Spec.ClusterIPs[0], "may not be set to 'None' for NodePort services"))
 		}
 	case core.ServiceTypeExternalName:
-		if service.Spec.ClusterIP != "" {
-			allErrs = append(allErrs, field.Forbidden(specPath.Child("clusterIP"), "must be empty for ExternalName services"))
+		// must have ClusterIP == "" && len(.spec.ClusterIPs == 0)
+		if len(service.Spec.ClusterIPs) > 0 {
+			allErrs = append(allErrs, field.Forbidden(specPath.Child("clusterIPs"), "may not be set for ExternalName services"))
+		}
+
+		// must have nil families and nil policy
```

Comment:
This breaks clients.  E.g. If they have an old client and:

* Create a service type=ClusterIP
* silently Get assigned ipFamiles and policy
* Change type to ExternalName

That sequence works today (without dual-stack) and would break when dual-stack is enabled.

This becomes roughly the same as NodePort and ClusterIP, but worse because it is breaking, so ignoring it is not really an answer

Options:
1) Clear them unilaterally whenever type=ExternalName
2) Clear them whenever changing to type=ExternalName
3) Clear them if we know we assigned them internally (harder, but @lavalamp thinks it is more correct - https://github.com/kubernetes/kubernetes/issues/95162)

3a) apimachinery gives us a way to answer "who owns this field"
3b) We parse managedFields ourselves
3c) We manage a new annotation for this
3d) We make a new resource to track this metadata

Reply:
The problem here is, contrary to other service types, that ExternalName does not have any IP family notion, is just a CNAME record.
I vote 1 because of that, and because I could not find an example that cleaning the fields for this type of service can break someone (but I may be totally wrong here)

Reply:
re #95162 works on managed fields. These new fields are not managed fields. or are they?

Reply:
this has been resolved by moving conversion and field clearing to strategy.